nssexec.c

```diff
@@ -0,0 +1,290 @@
+	 * or programming issue.
+	 */
+	if (!app)
+		err(EXIT_FAILURE, _("mapping tool not present"));
+
+	child = fork();
+	if (child < 0)
+	err(EXIT_FAILURE, _("failed to fork"));
+
+	if (!child) {
+#define MAX_ARGV 9
+		char *argv[MAX_ARGV];
+		char *envp[] = { NULL };
+		char pid_fmt[16];
+		int argc = 0;
+		unsigned start, length;
+
+		snprintf(pid_fmt, 16, "%d", pid);
+
+		argv[argc++] = (char *)app;
+		argv[argc++] = pid_fmt;
+		argv[argc++] = "0";
+		argv[argc++] = (char *)id;
+		argv[argc++] = "1";
+		argv[argc++] = "1";
+	/*
+	 * Convert the map string into a list of argument that
+	 * newuidmap/newgidmap can understand.
+	 */
+
+		while ((range = range_item(range, &start, &length))) {
+			char startstr[16];
+			char lengthstr[16];
+			sprintf(startstr, "%u", start);
+			sprintf(lengthstr, "%u", length);
+			argv[6] = startstr;
+			argv[7] = lengthstr;
+			argv[8]= (char*)0;
+		}
+
+		int i;
+		for(i=0;i<MAX_ARGV;i++)
+		{
+			printf("%s ",argv[i]);
+		}
+		printf("\n");
+
+		execve(app, argv, envp);
+		fflush(stdout);
+		fflush(stderr);
+		err(EXIT_FAILURE, _("failed to execv"));
+	} else {
+		int status;
+
+		while (1) {
+			if (waitpid(child, &status, 0) < 0) {
+				if (errno == EINTR)
+					continue;
+				err(EXIT_FAILURE, _("failed to waitpid"));
+			}
+			if (WIFEXITED(status) || WIFSIGNALED(status))
+				return WEXITSTATUS(status);
+		}
+	}
+
+	return -1;
+}
+
+static char *read_ranges(int type) {
+	char *line = NULL, *entry, *range, *user;
+	size_t end, size;
+	//struct passwd *passwd;
+	uid_t uid;
+	unsigned int length, start;
+	FILE *file;
+
+	range = string("%u:1", getid(type));
+	if (!(file = fopen(subpath(type), "r")))
+		return range;
+
+	uid = getuid();
+	user = getenv("USER");
+	user = user ? user : getlogin();
+
+	while (getline(&line, &size, file) >= 0) {
+		if (strtol(line, &entry, 10) != uid || entry == line) {
+			if (strncmp(line, user, strlen(user)))
+				continue;
+			entry = line + strlen(user);
+		}
+		if (sscanf(entry, ":%u:%u%zn", &start, &length, &end) < 2)
+			continue;
+		if (strchr(":\n", entry[end + 1]))
+			append(&range, ",%u:%u", start, length);
+	}
+
+	free(line);
+	fclose(file);
+
+	return range;
+}
+
+static void set_propagation(unsigned long flags)
+{
+	if (flags == 0)
+		return;
+
+	if (mount("none", "/", NULL, flags, NULL) != 0)
+		err(EXIT_FAILURE, _("cannot change root filesystem propagation"));
+}
+
+void nsexec(void)
+{
+	pid_t pid = 0;
+	unsigned long propagation = UNSHARE_PROPAGATION_DEFAULT;
+
+	uid_t real_euid = geteuid();
+	gid_t real_egid = getegid();
+
+	char euid_fmt[16];
+	char egid_fmt[16];
+	snprintf(euid_fmt, 16, "%u",real_euid);
+	snprintf(egid_fmt, 16, "%u",real_egid);
+
+	char *uid_map;
+	uid_map = read_ranges(UID);
+	char *gid_map;
+	gid_map = read_ranges(GID);
+
+	if (unshare(CLONE_NEWNS | CLONE_NEWUSER) == -1){
+		err(EXIT_FAILURE, _("unshare failed"));
+	}
+
+	set_propagation(propagation);
+
+	pid = getpid();
+
+	if (try_mapping_tool(idtool(UID), pid, uid_map, euid_fmt))
```

Comment:
These need to be run from the host namespace -- you can't run them from inside the namespace (since `setuid` binaries won't do what you expect inside a user namespace). In `runc` we `fork()` multiple times and so have a parent that does the mappings for the final `runc init`.